working/base-interface-final/feature-specification.md

```diff
@@ -0,0 +1,470 @@
+*   `final` and `interface` imply that the type can't be used as a superclass,
+    but the reason for defining a mixin is so that it can be mixed in and become
+    a superclass, so we don't allow those on mixins.
+
+The remaining valid combinations are:
+
+```
+class
+sealed class
+base class
+interface class
+final class
+abstract class
+abstract base class
+abstract interface class
+abstract final class
+
+mixin
+sealed mixin
+base mixin
+interface mixin
+final mixin
+```
+
+The grammar is:
+
+```
+classDeclaration ::=
+  classModifiers 'class' identifier typeParameters?
+  superclass? interfaces?
+  '{' (metadata classMemberDeclaration)* '}'
+  | classModifiers 'class' mixinApplicationClass
+
+mixinDeclaration ::= mixinModifiers 'mixin' identifier typeParameters?
+  ('on' typeNotVoidList)? interfaces?
+  '{' (metadata classMemberDeclaration)* '}'
+
+classModifiers ::= 'sealed' | 'abstract'? typeModifiers?
+mixinModifiers ::= 'sealed' | typeModifiers?
+typeModifiers  ::= 'base' | 'interface' | 'final'
+```
+
+### Static semantics
+
+It is a compile-time error to:
+
+*   Extend a class marked `interface` or `final` outside of the library where it
+    is defined.
+
+*   Implement a type marked `base` or `final` outside of the library where it is
+    defined.
+
+*   Extend or mix in a type marked `base` outside of the library where it is
+    defined without also being marked `base` or `final`. *This ensures that a
+    subtype can't escape the `base` restriction of its supertype by offering its
+    _own_ interface that could then be implemented without inheriting the
+    concrete implementation from the supertype.*
+
+*   Mix in a class marked `sealed`, `base`, `interface`, or `final`. *We want to
+    eventually move away from classes as mixins. We don't want to break existing
+    uses of classes as mixins but since no existing code is using these
+    modifiers, we can prevent classes using those modifiers from also being used
+    as mixins.*
+
+A typedef can't be used to subvert these restrictions. When extending,
+implementing, or mixing in a typedef, we look at the library where type the
+typedef resolves to is defined to determine if the behavior is allowed. *Note
+that the library where the _typedef_ is defined does not come into play.*
+
+### `@reopen` lint
+
+We don't specify lints and metadata annotations in the language specification,
+so this part of the proposal will not become a formal part of the language.
+Instead, it's a suggested part of the overall user experience of the feature.
+
+A metadata annotation `@reopen` is added to package [meta][] and a lint
+"require_reopen" is added to the [linter][]. When the lint is enabled, a lint
+warning is reported if a class or mixin is not annotated `@reopen` and it:
+
+*   Extends or mixes in a type marked `interface` or `final` and is not itself
+    marked `interface` or `final`.
+
+*   Extends, implements, or mixes in a type marked `base` or `final` and is
```

Comment:
Is this directly extends, implements or mixes in, or transitive?

If transitive, we need to be concerned about paths going through other reopened classes, possibly through other libraries. If not transitive ...

Reply:
I think directly.

Reply:
I haven't thought deeply about it, but I don't think it should be transitive? If you've already deliberately re-added a capability and put `@reopen` on that the type that does it, at that point further subtypes of the reopened type shouldn't need to restate that they also have the capability.